src/cargo/ops/cargo_install.rs

```diff
@@ -0,0 +1,329 @@
+            });
+            let examples = candidates.iter().filter(|cand| {
+                cand.targets().iter().filter(|t| t.is_example()).count() > 0
+            });
+            let pkg = match try!(one(binaries, |v| multi_err("binaries", v))) {
+                Some(p) => p,
+                None => {
+                    match try!(one(examples, |v| multi_err("examples", v))) {
+                        Some(p) => p,
+                        None => return Err(human("no packages found with \
+                                                  binaries or examples")),
+                    }
+                }
+            };
+            return Ok((pkg.clone(), Box::new(source)));
+
+            fn multi_err(kind: &str, mut pkgs: Vec<&Package>) -> String {
+                pkgs.sort_by(|a, b| a.name().cmp(b.name()));
+                format!("multiple packages with {} found: {}", kind,
+                        pkgs.iter().map(|p| p.name()).collect::<Vec<_>>()
+                            .join(", "))
+            }
+        }
+    }
+}
+
+fn one<I, F>(mut i: I, f: F) -> CargoResult<Option<I::Item>>
+    where I: Iterator,
+          F: FnOnce(Vec<I::Item>) -> String
+{
+    match (i.next(), i.next()) {
+        (Some(i1), Some(i2)) => {
+            let mut v = vec![i1, i2];
+            v.extend(i);
+            Err(human(f(v)))
+        }
+        (Some(i), None) => Ok(Some(i)),
+        (None, _) => Ok(None)
+    }
+}
+
+fn check_overwrites(dst: &Path,
+                    pkg: &Package,
+                    filter: &ops::CompileFilter,
+                    prev: &CrateListingV1) -> CargoResult<()> {
+    let check = |name| {
+        let name = format!("{}{}", name, env::consts::EXE_SUFFIX);
+        if fs::metadata(dst.join(&name)).is_err() {
+            return Ok(())
+        }
+        let mut msg = format!("binary `{}` already exists in destination", name);
+        if let Some((p, _)) = prev.v1.iter().find(|&(_, v)| v.contains(&name)) {
+            msg.push_str(&format!(" as part of `{}`", p));
+        }
+        Err(human(msg))
+    };
+    match *filter {
+        CompileFilter::Everything => {
+            // If explicit --bin or --example flags were passed then those'll
+            // get checked during cargo_compile, we only care about the "build
+            // everything" case here
+            if pkg.targets().iter().filter(|t| t.is_bin()).next().is_none() {
+                return Err(human("specified package has no binaries"))
+            }
+
+            for target in pkg.targets().iter().filter(|t| t.is_bin()) {
+                try!(check(target.name()));
+            }
+        }
+        CompileFilter::Only { bins, examples, .. } => {
+            for bin in bins.iter().chain(examples) {
+                try!(check(bin));
+            }
+        }
+    }
+    Ok(())
+}
+
+fn read_crate_list(path: &Path) -> CargoResult<CrateListingV1> {
+    let metadata = path.join(".crates.toml");
```

Comment:
If I'm reading this correctly, this has _one_ metadata file per install destination, listing all crates installed there (with their inner metadata)?

If so, this is *exactly* the kind of design I mentioned as being actively hostile to system package managers, as crate installs now mutate a single shared state, rather than being separable by way of installing individual drop-in files to a directory. I really dislike this, as it forms an ABI which (even if not public) will be a pain to change later - and if not changed, makes interop with system PMs needlessly difficult and fraught.